ert_shared/record.py

```diff
@@ -0,0 +1,184 @@
+from abc import abstractmethod, abstractclassmethod
+import asyncio
+from pathlib import Path
+from typing import Awaitable, Dict, Union, Mapping, Tuple, List
+
+from pydantic import BaseModel, root_validator
+
+from enum import Enum, auto
+
+from ert_shared.ensemble_evaluator.prefect_ensemble.storage_driver import _SharedDiskStorageDriver
+
+class RecordType(Enum):
+    matrix = auto()
+    vector = auto()
+    float_vector = auto()
+    int_vector = auto()
+    opaque = auto()
+
+
+class _DataElement(BaseModel):
+    class Config:
+        validate_all = True
+        validate_assignment = True
+        extra = "forbid"
+        allow_mutation = False
+        arbitrary_types_allowed = True
+
+
+def _build_record_index(data):
+    if isinstance(data, Mapping):
+        return tuple(data.keys())
+    else:
+        return tuple(range(len(data)))
+
+
+def _infer_mime_from_type(type: RecordType) -> str:
+    return "text/csv"
+
+
+class Record(_DataElement):
```

Comment:
This is a bit interesting because we are now in the scenario where we need shared code between the ensemble evaluator and the ert3 implementation. Do you have a strategy for how to organise this? Currently this appears to be an extended copy of the record implementation in `ert3.data`...